data/virt_autotest/fetch_logs_from_guest.sh

```diff
@@ -0,0 +1,260 @@
+            echo -e "Failed to fetch ${logs_folder} from guest ${guest_domain} via ssh."
+        fi
+
+	return ${ret_result}
+}
+
+#Fetch logs from virtual machine to local host by using libguestfs tools
+function fetch_logs_from_guest_via_libguestfS() {
+        local guest_domain=$1
+	local logs_fetched=$2
+	local logs_folder=$3
+
+        setup_libguestfs_env
+	virt-filesystems -d ${guest_domain} &> /dev/null
+	if [[ $? -ne 0 ]];then
+	   echo -e "Running ${guest_domain} can not be accessed by libguestfs currently. Shut it down now."
+	   virsh destroy ${guest_domain}
+	fi
+	local guest_device=`find_disk_hosts_filesystem ${guest_domain} ${logs_fetched}`
+        echo -e "${guest_domain} ${guest_device} contains ${logs_fetched}."
+        local guest_transformed=${guest_domain//./_}
+        mkdir -p ${logs_folder}/${guest_transformed}
+        echo -e "guestfish -r -d ${guest_domain} -m ${guest_device} copy-out ${logs_fetched} ${logs_folder}/${guest_transformed}"
+        guestfish -r -d ${guest_domain} -m ${guest_device} copy-out ${logs_fetched} ${logs_folder}/${guest_transformed}
+        if [[ $? -eq 0 ]];then
+           echo -e "Copied out ${logs_fetched} from ${guest_domain} successfully."
+           guestfish -r -d ${guest_domain} -m ${guest_device} rm-rf ${logs_fetched}
+           return 0
+        else
+           echo -e "Try again to mount ${guest_device}:/:subvol=@ with guestfish."
+           echo -e "guestfish -r -d ${guest_domain} -m ${guest_device}:/:subvol=@ copy-out ${logs_fetched} ${logs_folder}/${guest_transformed}"
+           guestfish -r -d ${guest_domain} -m ${guest_device}:/:subvol=@ copy-out ${logs_fetched} ${logs_folder}/${guest_transformed}
+           if [[ $? -eq 0 ]];then
+              echo -e "Copied out ${logs_fetched} from ${guest_domain} successfully."
+              guestfish -r -d ${guest_domain} -m ${guest_device}:/:subvol=@ rm-rf ${logs_fetched}
+              return 0
+           else
+              echo -e "Failed to copy out ${logs_fetched} from ${guest_domain}."
+              return 1
+           fi
+	fi
+}
+
+#Fetch logs from virtual machine to local host via ssh firstly. Will resort to libguestfs tools if ssh connection is broken
+function fetch_logs_from_guest() {
+	local guest_domain=$1
+	local guest_ipaddr=$2
+	local logs_folder=$3
+	shift
+	shift
+	shift
+	local extra_logs=($@)
+
+	local ret1=0
+	fetch_logs_from_guest_via_ssh ${guest_domain} ${guest_ipaddr} ${logs_folder}
+	if [[ $? -ne 0 ]];then
+	   echo -e "Try to use libguestfs tools to fetch ${logs_folder} from ${guest_domain}."
+           fetch_logs_from_guest_via_libguestfS ${guest_domain} ${logs_folder} ${logs_folder}
+	   ret1=$?
+	fi
+
+	local ret2=0
+	if [[ ${extra_logs[@]} != "" ]];then
+           echo -e "Try to use libguestfs tools to fetch ${extra_logs[@]} from ${guest_domain}."
+           for eachlog in ${extra_logs[@]};do
+	       fetch_logs_from_guest_via_libguestfS ${guest_domain} ${eachlog} ${logs_folder}
```

Comment:
Why not try fetch_logs_from_guest_via_ssh firstly for extra logs? If the above fetch_logs_from_guest_via_ssh is successful for ${logs_folder}, then the extra logs should use ssh first also(to avoid destroying the guest state(on->off)), right?

BTW, why not treat ${logs_folder} and ${extra_logs[@]} together in one loop, fetch_logs_from_guest_via_ssh first, then fetch_logs_from_guest_via_libguestfS ?

Reply:
1. Because extra logs from guest collecting via ssh already can happen in virt_logs_collector.sh. If that is successful, they are already in guest logs folder. If that is not successful, this script can handle fetch them out directly to host. 

2. Because guestfish copy-out does not support copy mutiple files or folder at the same time, so can only be done one by one.

Reply:
> 1. Because extra logs from guest collecting via ssh already can happen in virt_logs_collector.sh. If that is successful, they are already in guest logs folder. If that is not successful, this script can handle fetch them out directly to host.

Do you mean line 126, fetch_logs_from_guest_via_ssh ${guest_domain} ${guest_ipaddr} ${logs_folder}, without explicitly giving the extra logs, it already tries to fetch the extra logs via ssh? But I do not see it from code. Can you point which line does this?

> 2. Because guestfish copy-out does not support copy mutiple files or folder at the same time, so can only be done one by one.

Sure.

Seems you missed my point. If I understand correctly, what you are trying to do here is to first fetch guest supportconfig log, then extra guest logs. Both of them should try ssh first, if not work, then console way, right? But currect code logic seems to miss the ssh trying for extra logs because when ret1=0 after fetch_logs_from_guest_via_ssh, you directly use fetch_logs_from_guest_via_libguestfS, rather than first try fetch_logs_from_guest_via_ssh. So I suggest sth like below:

for log in  [ ${logs_folder} ${extra_logs[@]} ];do
   fetch_logs_from_guest_via_ssh
   if [$? -ne 0];then
       fetch_logs_from_guest_via_libguestfS
   fi
done

Reply:
Because you can not fetch logs out of guest to host via console. If ssh does not work, console can not help you out on fetching something out of guest.

Reply:
Sorry, my typo, console->libguestfs. The idea is still the same.

If I understand correctly, what you are trying to do here is to first fetch guest supportconfig log, then extra guest logs. Both of them should try ssh first, if not work, then libguestfs way, right? But currect code logic seems to miss the ssh trying for extra logs because when ret1=0 after fetch_logs_from_guest_via_ssh, you directly use fetch_logs_from_guest_via_libguestfS, rather than first try fetch_logs_from_guest_via_ssh. So I suggest sth like below:

for log in [ ${logs_folder} ${extra_logs[@]} ];do
  fetch_logs_from_guest_via_ssh
  if [$? -ne 0];then
    fetch_logs_from_guest_via_libguestfS
  fi
done

Reply:
I already told you it is already been done in function collect_extra_logs_from_guest in virt_logs_collector.sh script. This script is responsible for fetching, which means fetching everything we need to host logs folder then compress altogether. Collecting logs happens in virt_logs_collector.sh script.

Reply:
I see now. I did misunderstand before. Sorry about that. I am starting to curious about how in openqa postfail hook will you handle the various log collection  and log fetch return result now.